---
title: "TCR analysis"
author: "Kaushik Amancherla"
date: "2024-05-03"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(Seurat)
library(scRepertoire)
library(psych)

# Set working directory
setwd("C:/Users/amanchk/Downloads/PBMC_CAV_2.2024")
```

## Load PBMC CAV Seurat object

Here, I am just loading the object that I previously created. Once loaded, I will change name of two columns (cluster and subcluster) to make downstream analyses a bit easier.

```{r}
# Load Seurat object
pbmc <- readRDS("Processing_RDS_files/2024_pbmc_ALL_samples_postANNOTATION.rds")

# Rename two metadata columns for easier downstream analyses
pbmc@meta.data <- pbmc@meta.data %>% rename(
  cluster = predicted.celltype.l1,
  subcluster = predicted.celltype.l2
)

# Clear memory
gc()
```

## Load TRUST4 results
These are files generated by TRUST4 - TCR and BCR sequences. The BAM files from the original CellRanger outputs of each of the 8 multiplexed samples were used (labeled batch_1, batch_2, batch_3, etc. to correspond to each multiplexed sample -- e.g., batch_1 corresponds to 8256-SB-1, batch_2 corresponds to 8256-SB-2, etc.).

I will use scRepertoire to load them. I am using the development version, 2.0.3, of scRepertoire.

```{r}
# Make folder
folder <- "TRUST4/"

contig.list <- loadContigs(folder, "TRUST4")
head(contig.list[[1]])
```

I want to add the meta data from the single-cell dataset to these samples - I need to use batch and barcode as the anchors. When I was doing the scRNA-seq analysis, I made a separate barcode column in the metadata so that I do not have to worry about the changes that Seurat makes during the QC/normalization/clustering process.

I will first rename the items in the list (contig.list) by batch. Then, using the batch and barcode, I will merge metadata from the pbmc scRNA-seq RDS.
```{r}
# Metadata file
pbmc.metadata <- pbmc@meta.data
table(pbmc.metadata$batch)

# Relabel list items based on batch
contig.list <- set_names(contig.list, "batch_1",
                         "batch_2",
                         "batch_3",
                         "batch_4",
                         "batch_5",
                         "batch_6",
                         "batch_7",
                         "batch_8")

# Add metadata to contig.list
contig.list <- map2(contig.list, names(contig.list), function(data, batch_name) {
  # Filter pbmc.metadata for the current batch
  meta_subset <- filter(pbmc.metadata, batch == batch_name)
  
  # Merge metadata into the data frame based on barcode
  enriched_data <- left_join(data, meta_subset, by = "barcode")
  
  return(enriched_data)
})

head(contig.list[[1]])
```

## Combining contigs into clones
Next, I will combine the contigs into TCR clones.

```{r}
combined.TCR <- combineTCR(contig.list)
head(combined.TCR[[1]])
```

Maybe I should combine the metadata now? Since it disappeared.

```{r}
# Relabel list items based on batch
combined.TCR <- set_names(combined.TCR, "batch_1",
                         "batch_2",
                         "batch_3",
                         "batch_4",
                         "batch_5",
                         "batch_6",
                         "batch_7",
                         "batch_8")

# Add metadata to combined.TCR
combined.TCR <- map2(combined.TCR, names(combined.TCR), function(data, batch_name) {
  # Filter pbmc.metadata for the current batch
  meta_subset <- filter(pbmc.metadata, batch == batch_name)
  
  # Merge metadata into the data frame based on barcode
  enriched_data <- left_join(data, meta_subset, by = "barcode")
  
  return(enriched_data)
})

head(combined.TCR[[1]])
```

There are some NAs in columns that contain metadata from the pbmc RDS file. This means that those represent "cells" that were filtered out during the scRNA-seq QC. These may just represent ambient RNA that was reconstructed as a TCR in TRUST4.

I will remove those NAs, based on the Sample_ID column:
```{r}
combined.TCR <- map(combined.TCR, ~ filter(.x, !is.na(Sample_ID)))
head(combined.TCR[[1]])
```

## TCR analyses

First, I will quantify clones across the CAV grades.
```{r}
# Load metadata file
meta <- read_csv("pbmc_metadata.csv")

clone_quant <- clonalQuant(combined.TCR,
            cloneCall = "strict",
            chain = "both",
            scale = TRUE,
            group.by = "Sample_ID")
clone_quant

# Extract table for statistical analysis
clone_quant_samples <- clone_quant$data
clone_quant_samples <- inner_join(clone_quant_samples, meta, by = "Sample_ID")
clone_quant_samples <- clone_quant_samples %>%
  mutate(
    cav_grade = case_when(cav_grade == 0 ~ "CAV-0",
                          cav_grade == 1 ~ "CAV-1",
                          cav_grade == 2 ~ "CAV-2",
                          cav_grade == 3 ~ "CAV-3")
  )

# Set levels
clone_quant_samples$cav_grade <- factor(clone_quant_samples$cav_grade,
                                        levels = c("CAV-0", "CAV-1", "CAV-2", "CAV-3"))
levels(clone_quant_samples$cav_grade)

# Statistical analysis
print(kruskal.test(contigs ~ cav_grade, data = clone_quant_samples))

# Re-do figure by CAV grade to save
clonalQuant(combined.TCR,
            cloneCall = "strict",
            chain = "both",
            scale = TRUE,
            group.by = "cav_grade") + 
  theme(legend.title = element_text("CAV Grade")) +
  xlab("CAV Grade") + NoLegend()

# Save figure
# ggsave("Figures/TCR_clonalQuant.png", dpi = 1200, bg = "white")
```

Next, I will evaluate the relative space occupied by clones at specific proportions.
```{r}

clone_prop <- clonalHomeostasis(combined.TCR, 
                  cloneCall = "gene",
                  group.by = "Sample_ID")
clone_prop

# Extract table for statistical analysis
clone_prop_samples <- clone_prop$data
clone_prop_samples <- clone_prop_samples %>%
  dplyr::rename(Sample_ID = Var1)

clone_prop_samples <- inner_join(clone_prop_samples, meta, by = "Sample_ID")
clone_prop_samples <- clone_prop_samples %>%
  mutate(
    cav_grade = case_when(cav_grade == 0 ~ "CAV-0",
                          cav_grade == 1 ~ "CAV-1",
                          cav_grade == 2 ~ "CAV-2",
                          cav_grade == 3 ~ "CAV-3")
  )

# Set levels
clone_prop_samples$cav_grade <- factor(clone_prop_samples$cav_grade,
                                        levels = c("CAV-0", "CAV-1", "CAV-2", "CAV-3"))
levels(clone_prop_samples$cav_grade)

# Break down by clone proportion category
rare <- clone_prop_samples %>% filter(Var2 == "Rare (0 < X <= 1e-04)")
small <- clone_prop_samples %>% filter(Var2 == "Small (1e-04 < X <= 0.001)")
medium <- clone_prop_samples %>% filter(Var2 == "Medium (0.001 < X <= 0.01)")
large <- clone_prop_samples %>% filter(Var2 == "Large (0.01 < X <= 0.1)")
hyper <- clone_prop_samples %>% filter(Var2 == "Hyperexpanded (0.1 < X <= 1)")

# Statistical analysis
print(kruskal.test(value ~ cav_grade, data = rare))
print(kruskal.test(value ~ cav_grade, data = small))
print(kruskal.test(value ~ cav_grade, data = medium))
print(kruskal.test(value ~ cav_grade, data = large))
print(kruskal.test(value ~ cav_grade, data = hyper))

# Figure for manuscript
clonalHomeostasis(combined.TCR, 
                  cloneCall = "gene",
                  group.by = "cav_grade") + 
  theme(legend.title = element_text("CAV Grade")) +
  xlab("CAV Grade")

# Save figure
# ggsave("Figures/TCR_clonalHomeostasis.png", dpi = 1200, bg = "white")
```
Finally, I will evaluate clonal diversity.
```{r}
clonal_diversity <- clonalDiversity(combined.TCR, 
                cloneCall = "gene",
                group.by = "Sample_ID")

# Extract table for statistical analysis
clone_diversity_samples <- clonal_diversity$data
clone_diversity_samples <- inner_join(clone_diversity_samples, meta, by = "Sample_ID")
clone_diversity_samples <- clone_diversity_samples %>%
  mutate(
    cav_grade = case_when(cav_grade == 0 ~ "CAV-0",
                          cav_grade == 1 ~ "CAV-1",
                          cav_grade == 2 ~ "CAV-2",
                          cav_grade == 3 ~ "CAV-3")
  )

# Set levels
clone_diversity_samples$cav_grade <- factor(clone_diversity_samples$cav_grade,
                                        levels = c("CAV-0", "CAV-1", "CAV-2", "CAV-3"))
levels(clone_diversity_samples$cav_grade)

# Break down by clone proportion category
shannon <- clone_diversity_samples %>% filter(variable == "shannon")
inv.simpson <- clone_diversity_samples %>% filter(variable == "inv.simpson")
norm.entropy <- clone_diversity_samples %>% filter(variable == "norm.entropy")
gini.simpson <- clone_diversity_samples %>% filter(variable == "gini.simpson")
chao1 <- clone_diversity_samples %>% filter(variable == "chao1")
ACE <- clone_diversity_samples %>% filter(variable == "ACE")
  
# Statistical analysis
print(kruskal.test(value ~ cav_grade, data = shannon))
print(kruskal.test(value ~ cav_grade, data = inv.simpson))
print(kruskal.test(value ~ cav_grade, data = norm.entropy))
print(kruskal.test(value ~ cav_grade, data = gini.simpson))
print(kruskal.test(value ~ cav_grade, data = chao1))
print(kruskal.test(value ~ cav_grade, data = ACE))

# Figure for manuscript
clonalDiversity(combined.TCR, 
                cloneCall = "gene",
                group.by = "cav_grade") +
  scale_fill_discrete(name="CAV Grade")

# Save figure
# ggsave("Figures/TCR_clonalDiversity.png", dpi = 1200, bg = "white", width = 12)
```
